stage2, translate-c: mod math with signed ints results in compile error

### Zig Version

0.10.0-dev.4634+c36eb4ede

### Steps to Reproduce and Observed Behavior

`translate-c` the following and use the resulting function

```c
#define GLAD_VERSION_MINOR(version) (version % 10000)
```

```zig
c.GLAD_VERSION_MINOR(@as(c_int, 20000));
```

```
cimport.zig:1372:60: error: remainder division with 'c_int' and 'c_int': signed integers and floats must use @rem or @mod
```

Note that if I use a `c_uint` it works fine, as expected.

### Expected Behavior

Works!

--- a/lib/std/zig/c_translation.zig
+++ b/lib/std/zig/c_translation.zig
@@ -570,6 +570,16 @@ pub const MacroArithmetic = struct {
             else => unreachable,
         }
     }
+
+    pub fn mod(a: anytype, b: anytype) ArithmeticConversion(@TypeOf(a), @TypeOf(b)) {
+        const ResType = ArithmeticConversion(@TypeOf(a), @TypeOf(b));
+        const a_casted = cast(ResType, a);
+        const b_casted = cast(ResType, b);
+        switch (@typeInfo(ResType)) {
+            .Int => return @mod(a_casted, b_casted),
+            else => unreachable,
+        }
+    }
 };
 
 test "Macro suffix functions" {

--- a/lib/std/zig/c_translation.zig
+++ b/lib/std/zig/c_translation.zig
@@ -571,7 +571,7 @@ pub const MacroArithmetic = struct {
         }
     }
 
-    pub fn mod(a: anytype, b: anytype) ArithmeticConversion(@TypeOf(a), @TypeOf(b)) {
+    pub fn rem(a: anytype, b: anytype) ArithmeticConversion(@TypeOf(a), @TypeOf(b)) {
         const ResType = ArithmeticConversion(@TypeOf(a), @TypeOf(b));
         const a_casted = cast(ResType, a);
         const b_casted = cast(ResType, b);

--- a/lib/std/zig/c_translation.zig
+++ b/lib/std/zig/c_translation.zig
@@ -576,7 +576,13 @@ pub const MacroArithmetic = struct {
         const a_casted = cast(ResType, a);
         const b_casted = cast(ResType, b);
         switch (@typeInfo(ResType)) {
-            .Int => return @mod(a_casted, b_casted),
+            .Int => {
+                if (@typeInfo(ResType).Int.signedness == .signed) {
+                    return signedRemainder(a_casted, b_casted);
+                } else {
+                    return a_casted % b_casted;
+                }
+            },
             else => unreachable,
         }
     }

--- a/src/translate_c.zig
+++ b/src/translate_c.zig
@@ -6237,7 +6237,7 @@ fn parseCMulExpr(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!Node {
             .Percent => {
                 const lhs = try macroBoolToInt(c, node);
                 const rhs = try macroBoolToInt(c, try parseCCastExpr(c, m, scope));
-                node = try Tag.mod.create(c.arena, .{ .lhs = lhs, .rhs = rhs });
+                node = try Tag.macro_arithmetic.create(c.arena, .{ .op = .mod, .lhs = lhs, .rhs = rhs });
             },
             else => {
                 m.i -= 1;

--- a/src/translate_c.zig
+++ b/src/translate_c.zig
@@ -6237,7 +6237,7 @@ fn parseCMulExpr(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!Node {
             .Percent => {
                 const lhs = try macroBoolToInt(c, node);
                 const rhs = try macroBoolToInt(c, try parseCCastExpr(c, m, scope));
-                node = try Tag.macro_arithmetic.create(c.arena, .{ .op = .mod, .lhs = lhs, .rhs = rhs });
+                node = try Tag.macro_arithmetic.create(c.arena, .{ .op = .rem, .lhs = lhs, .rhs = rhs });
             },
             else => {
                 m.i -= 1;

--- a/src/translate_c/ast.zig
+++ b/src/translate_c/ast.zig
@@ -726,7 +726,7 @@ pub const Payload = struct {
             rhs: Node,
         },
 
-        pub const Operator = enum { div, mod };
+        pub const Operator = enum { div, rem };
     };
 };
 

--- a/src/translate_c/ast.zig
+++ b/src/translate_c/ast.zig
@@ -726,9 +726,7 @@ pub const Payload = struct {
             rhs: Node,
         },
 
-        pub const Operator = enum {
-            div,
-        };
+        pub const Operator = enum { div, mod };
     };
 };
 

--- a/test/behavior/translate_c_macros.h
+++ b/test/behavior/translate_c_macros.h
@@ -57,5 +57,5 @@ typedef _Bool uintptr_t;
 #define DIVIDE_CONSTANT(version) (version / 1000)
 #define DIVIDE_ARGS(A, B) (A / B)
 
-#define MODULUS_CONSTANT(version) (version % 10000)
-#define MODULUS_ARGS(A, B) (A % B)
+#define REMAINDER_CONSTANT(version) (version % 10000)
+#define REMAINDER_ARGS(A, B) (A % B)

--- a/test/behavior/translate_c_macros.h
+++ b/test/behavior/translate_c_macros.h
@@ -56,3 +56,6 @@ typedef _Bool uintptr_t;
 
 #define DIVIDE_CONSTANT(version) (version / 1000)
 #define DIVIDE_ARGS(A, B) (A / B)
+
+#define MODULUS_CONSTANT(version) (version % 10000)
+#define MODULUS_ARGS(A, B) (A % B)

--- a/test/behavior/translate_c_macros.zig
+++ b/test/behavior/translate_c_macros.zig
@@ -172,18 +172,18 @@ test "Macro that uses division operator. Issue #13162" {
     );
 }
 
-test "Macro that uses modulus operator. Issue #13346" {
+test "Macro that uses remainder operator. Issue #13346" {
     if (builtin.zig_backend == .stage2_wasm) return error.SkipZigTest; // TODO
     if (builtin.zig_backend == .stage2_x86_64) return error.SkipZigTest; // TODO
     if (builtin.zig_backend == .stage2_arm) return error.SkipZigTest; // TODO
     if (builtin.zig_backend == .stage2_aarch64) return error.SkipZigTest; // TODO
 
-    try expectEqual(@as(c_int, 2_010), h.MODULUS_CONSTANT(@as(c_int, 42_010)));
-    try expectEqual(@as(c_uint, 2_030), h.MODULUS_CONSTANT(@as(c_uint, 42_030)));
+    try expectEqual(@as(c_int, 2_010), h.REMAINDER_CONSTANT(@as(c_int, 42_010)));
+    try expectEqual(@as(c_uint, 2_030), h.REMAINDER_CONSTANT(@as(c_uint, 42_030)));
 
     try expectEqual(
         @as(c_int, 7),
-        h.MODULUS_ARGS(
+        h.REMAINDER_ARGS(
             @as(i8, 17),
             @as(i8, 10),
         ),
@@ -191,7 +191,7 @@ test "Macro that uses modulus operator. Issue #13346" {
 
     try expectEqual(
         @as(c_int, 6),
-        h.MODULUS_ARGS(
+        h.REMAINDER_ARGS(
             @as(c_ushort, 25),
             @as(c_ushort, 20),
         ),

--- a/test/behavior/translate_c_macros.zig
+++ b/test/behavior/translate_c_macros.zig
@@ -171,3 +171,29 @@ test "Macro that uses division operator. Issue #13162" {
         ),
     );
 }
+
+test "Macro that uses modulus operator. Issue #13346" {
+    if (builtin.zig_backend == .stage2_wasm) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_x86_64) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_arm) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_aarch64) return error.SkipZigTest; // TODO
+
+    try expectEqual(@as(c_int, 2_010), h.MODULUS_CONSTANT(@as(c_int, 42_010)));
+    try expectEqual(@as(c_uint, 2_030), h.MODULUS_CONSTANT(@as(c_uint, 42_030)));
+
+    try expectEqual(
+        @as(c_int, 7),
+        h.MODULUS_ARGS(
+            @as(i8, 17),
+            @as(i8, 10),
+        ),
+    );
+
+    try expectEqual(
+        @as(c_int, 6),
+        h.MODULUS_ARGS(
+            @as(c_ushort, 25),
+            @as(c_ushort, 20),
+        ),
+    );
+}

--- a/test/behavior/translate_c_macros.zig
+++ b/test/behavior/translate_c_macros.zig
@@ -190,10 +190,18 @@ test "Macro that uses remainder operator. Issue #13346" {
     );
 
     try expectEqual(
-        @as(c_int, 6),
+        @as(c_int, 5),
         h.REMAINDER_ARGS(
             @as(c_ushort, 25),
             @as(c_ushort, 20),
+        ),
+    );
+
+    try expectEqual(
+        @as(c_int, 1),
+        h.REMAINDER_ARGS(
+            @as(c_int, 5),
+            @as(c_int, -2),
         ),
     );
 }

--- a/test/translate_c.zig
+++ b/test/translate_c.zig
@@ -511,7 +511,7 @@ pub fn addCases(cases: *tests.TranslateCContext) void {
         \\        _ = @as(c_int, 4) == @as(c_int, 4);
         \\        _ = @as(c_int, 5) * @as(c_int, 6);
         \\        _ = baz(@as(c_int, 1), @as(c_int, 2));
-        \\        _ = @as(c_int, 2) % @as(c_int, 2);
+        \\        _ = @import("std").zig.c_translation.MacroArithmetic.rem(@as(c_int, 2), @as(c_int, 2));
         \\        break :blk_1 baz(@as(c_int, 1), @as(c_int, 2));
         \\    };
         \\}

